app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java

```diff
@@ -967,52 +1179,71 @@ public synchronized boolean connect(final String address) {
         }
         Log.i(TAG, "connect: Trying to create a new connection.");
         setRetryTimer();
-        mBluetoothGatt = device.connectGatt(getApplicationContext(), true, mGattCallback);
+        mBluetoothGatt = device.connectGatt(getApplicationContext(), false, mGattCallback);
         mConnectionState = STATE_CONNECTING;
+        last_connect_request = JoH.tsl();
         return true;
     }
 
-    public void close() {
+    public synchronized void close() {
         Log.i(TAG, "close: Closing Connection - setting state DISCONNECTED");
         if (mBluetoothGatt == null) {
-            return;
-        }
-        try {
-            mBluetoothGatt.close();
-        } catch (NullPointerException e) {
-            Log.wtf(TAG, "Concurrency related null pointer in close");
+            Log.i(TAG, "not closing as bluetooth gatt is null");
+            //  return;
+        } else {
+            // TODO gatt refresh??
+
+            try {
+                mBluetoothGatt.close();
+            } catch (NullPointerException e) {
+                Log.wtf(TAG, "Concurrency related null pointer in close");
+            }
+
         }
 
         setRetryTimer();
         mBluetoothGatt = null;
         mCharacteristic = null;
         mConnectionState = STATE_DISCONNECTED;
         servicesDiscovered = DISCOVERED.NULL;
+        last_connect_request = 0;
+    }
+
+    private void sendReply(BridgeResponse reply) {
+        for (ByteBuffer byteBuffer : reply.send) {
+            Log.d(TAG, "Sending reply message");
+            sendBtMessage(byteBuffer);
+        }
     }
 
     public synchronized void setSerialDataToTransmitterRawData(byte[] buffer, int len) {
-        long timestamp = new Date().getTime();
-        last_time_seen = JoH.ts();
+
+        last_time_seen = JoH.tsl();
         watchdog_count = 0;
         if (static_use_blukon && Blukon.checkBlukonPacket(buffer)) {
             final byte[] reply = Blukon.decodeBlukonPacket(buffer);
             if (reply != null) {
                 Log.d(TAG, "Sending reply message from Blukon decoder");
                 sendBtMessage(reply);
             }
-        } else  if (blueReader.isblueReader()) {
+        } else if (blueReader.isblueReader()) {
             final byte[] reply = blueReader.decodeblueReaderPacket(buffer, len);
             if (reply != null) {
                 Log.d(TAG, "Sending reply message from blueReader decoder");
                 sendBtMessage(reply);
             }
-        } else  if (Tomato.isTomato()) {
-            ArrayList<ByteBuffer> buffers = Tomato.decodeTomatoPacket(buffer, len);
-            for (ByteBuffer byteBuffer : buffers ) {
-                Log.d(TAG, "Sending reply message from tomato decoder");
-                sendBtMessage(byteBuffer);
+        } else if (Tomato.isTomato()) {
```

Comment:
Here converted to use `BridgeResponse` instead of array list of bytebuffer and able to send reply with delay to avoid cross-talk when Tomato is sending us a stream of data.

Error messages passed to the status UI elements.